feature: add StaticProvider

peppsac opened the pull request:
Sometimes it's handy to be able to use a single image as a layer (mainly want you
want to quick test without the hassle of setting up a wms server for instance).

This provider is really simple, and just compute the offset in the image to use
to match the tile extent.

Example usage:
```js
    view.addLayer({
        url: 'http://localhost:8080/foo.png',
        networkOptions: { crossOrigin: 'anonymous' },
        type: 'color',
        protocol: 'image',
        id: 'myimage',
    });
```

autra commented on src/Core/Scheduler/Providers/SingleImageProvider.js:
```diff
@@ -0,0 +1,52 @@
+
+    getColorTexture(tile, layer) {
+        if (!this.tileInsideLimit(tile, layer)) {
+            return Promise.reject(`Tile '${tile}' is outside layer bbox ${layer.extent}`);
```
This will produce something like 

> Tile '[object Object]' is outside layer bbox '[object Object]'

Also, usually, people reject with an Error object, not a single string, which could allow you to attach tile and layer to this error object maybe? Not sure what is best here.

autra reviewed (commented):
Is it supposed to work with the globe? Is the following result expected?

![selection_038](https://user-images.githubusercontent.com/1576598/29785278-43096eda-8c27-11e7-8e82-b6a66e6cfd9f.png)

peppsac commented:
I'm going to refactor this one to support another use case: instead of using a single image, the user could give this provider a list of images and their respective extent.
This would make it easier to use itowns without the hassle of setting up an imagery server.

peppsac committed:
feature: add StaticProvider to download from a list of predefined images

Allows to use iTowns features without the trouble of setting up a WM(T)S server.

Given a metadata json file like this:

{
  'file1.png': [west1, east1, south1, north1],
  'file2.png': [west2, east2, south2, north2],
  ...
  'fileN.png': [westN, eastN, southN, northN]
}

The StaticProvider will simply find the best image for a given tile (= the image
with the smallest extent covering all of the tile's extent).

peppsac committed:
feature: add UPDATE_STATE.FINISHED state = no need to update this layer anymore

peppsac commented:
PR updated (depends on PR #485 )

I've refactored the code a bit and now this provider takes a list of image with their respective extent.

Note: this provider only makes sense for planar tiles.

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+            } else {
+                const d = selection.extent.dimensions();
+                const e = entry.extent.dimensions();
+                if (e.x <= d.x && e.y <= d.y) {
```
Is it possible to have a e.x > d.x but e.y < d.y? In this case, maybe we should choose the smallest one.

And btw, shouldn't we discriminate on the resulting px per meter instead? That's how I'd define "best" intuitively...

autra replied:
For my second point, I realised you can't do that before downloading the image (at least not without additional metadata) so we can forget it.

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        return currentTexture.file != s.image;
+    },
+
+    getColorTexture(tile, layer) {
```
I didn't see where you're using it elsewhere. If you're not, let's not export this.

peppsac replied:
Correct. I'll fix this.

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        layer.url = new URL(layer.url, window.location);
+        return Fetcher.json(layer.url.href).then((metadata) => {
+            layer.images = [];
+            // eslint-disable-next-line guard-for-in
```
eslint complains because you didn't check `if (metadata.hasOwnProperty(image))` inside the for-loop.

Note that you *might* need to use `Object.prototype.hasOwnProperty.call(metadata, image)` instead if it does not work, see https://eslint.org/docs/rules/guard-for-in

peppsac replied:
The help page isn't helpful: `This rule is aimed at preventing unexpected behavior that could arise from using a for in loop` : Sure, but what "unexpected behavior" could possibly happen? (especially in this case where metadata is a json file)?

autra replied:
Yes, in this particular case, we don't really risk anything. It's just that I prefer to add a guard than to disable the linter. But I should really have prefixed this with `nits`, it's not very important.

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        if (!this.tileInsideLimit(tile, layer)) {
+            return Promise.reject(`Tile '${tile}' is outside layer bbox ${layer.extent}`);
+        }
+        if (tile.material === null) {
```
maybe we can use `==` to also cover undefined? Or alternatively, as `tile.material` won't ever be 0, just `if (tile.material)`?

peppsac replied:
Yep, thanks.

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        const selection = selectBestImageForExtent(layer.images, tile.extent);
+
+        if (!selection) {
+            return Promise.reject('pas dimage');
```
Let's reject with a `Error` object (and let's translate that ;-) )

peppsac replied:
Oops 👍

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        }
+        const mat = tile.material;
+        const idx = mat.getLayerTextureOffset(layer.id);
+        const currentTexture = mat.textures[1][idx];
```
`mat.textures[l_COLOR][idx]`

autra commented on src/Core/Scheduler/Providers/StaticProvider.js:
```diff
@@ -0,0 +1,130 @@
+        const mat = tile.material;
+        const idx = mat.getLayerTextureOffset(layer.id);
+        const currentTexture = mat.textures[1][idx];
+        if (!currentTexture.file) {
```
Does this mean we have done the request but haven't received the response yet?

autra replied:
Actually, I'm not sure this situation is possible. Can you elaborate?

peppsac replied:
Textures are always not null in iTowns (see https://github.com/iTowns/itowns/blob/master/src/Renderer/LayeredMaterial.js#L133).
So in this provider I'm storing the image name  in an additional property (`file`) to be able to compare this filename to the best choice available.
It's the most naive way to implement this but it works.

autra reviewed (changes requested):
Nice! This will be useful indeed.

Can you provide a simple example with this PR?

peppsac commented:
PR updated.
I don't have any data to build a public example yet, but I'll open a PR soon to add an example.
